File: frontend/react-app/src/services/voiceAnalysis.js
// Lightweight client-side voice analysis with optional backend fallback
// Exports analyzeTranscript(transcript, durationSeconds)

export async function analyzeTranscript(transcript = '', durationSeconds = 0) {
  // Basic local analysis: count filler words and compute speaking speed
  const fillers = ['um', 'uh', 'like', 'you know', 'so', 'actually']
  const text = (transcript || '').toLowerCase()
  const words = text.trim().split(/\s+/).filter(Boolean)
  const wordCount = words.length

  let fillerCount = 0
  for (const f of fillers) {
    const re = new RegExp('\\b' + f.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&') + '\\b', 'g')
    const m = text.match(re)
    if (m) fillerCount += m.length
  }

  const speakingSpeed = durationSeconds > 0 ? (wordCount / Math.max(0.1, durationSeconds)) : 0
  const fillerRatio = wordCount > 0 ? (fillerCount / wordCount) : 0

  // confidence_score is heuristic: based on length and filler ratio
  const confidenceScore = Math.max(0, 100 * Math.min(1, (1 - fillerRatio) * Math.min(1, wordCount / 50)))

  return {
    filler_ratio: Number(fillerRatio.toFixed(3)),
    speaking_speed: Number(speakingSpeed.toFixed(3)),
    confidence_score: Number(confidenceScore.toFixed(1))
  }
}

export default { analyzeTranscript }

File: backend/utils/voice_analysis.py
"""Simple voice analysis utilities for backend (transcript-based).
These functions are intentionally lightweight and deterministic.
"""
import re

FILLERS = [r'\bum\b', r'\buh\b', r"\blike\b", r"\byou know\b", r"\bso\b", r"\bactually\b"]

def analyze_transcript(transcript: str, duration_seconds: float = 0.0):
    text = (transcript or '').lower()
    words = re.findall(r"\w+'?\w*|\w+", text)
    word_count = len(words)

    filler_count = 0
    for pat in FILLERS:
        filler_count += len(re.findall(pat, text))

    filler_ratio = (filler_count / word_count) if word_count else 0.0
    speaking_speed = (word_count / max(0.1, duration_seconds)) if duration_seconds else 0.0

    confidence_score = max(0.0, min(1.0, (1.0 - filler_ratio) * min(1.0, word_count / 50.0)))

    return {
        'filler_ratio': round(filler_ratio, 3),
        'speaking_speed': round(speaking_speed, 3),
        'confidence_score': round(confidence_score, 3)
    }
